File: android/app/src/main/kotlin/com/goalmaker/app/ui/review/ReviewViewModel.kt
package com.goalmaker.app.ui.review

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.goalmaker.app.application.planning.AreaList
import com.goalmaker.app.application.planning.GoalDraft
import com.goalmaker.app.application.planning.GoalList
import com.goalmaker.app.application.planning.GoalRules
import com.goalmaker.app.application.planning.HabitList
import com.goalmaker.app.application.planning.PlanDecision
import com.goalmaker.app.application.planning.Reflection
import com.goalmaker.app.application.planning.ReviewItem
import com.goalmaker.app.application.planning.ReviewList
import com.goalmaker.app.application.planning.ReviewLookBack
import com.goalmaker.app.application.planning.RitualRunList
import com.goalmaker.app.application.planning.TaskItem
import com.goalmaker.app.application.planning.TaskList
import com.goalmaker.app.domain.planning.PromptLibrary
import com.goalmaker.app.domain.planning.PromptRules
import com.goalmaker.app.domain.planning.ReviewQuestion
import com.goalmaker.app.ui.goals.GoalBoard
import java.time.LocalDate
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * The guided review of one period (docs/reviews.md, spec stories 59 to 64): look back, handle what is
 * still open, reflect on the prompts the library and the period's facts give, rate mood and energy,
 * and set the next period's goals. Everything is saved as it is answered.
 */
class ReviewViewModel(
    private val kind: String,
    private val periodStart: LocalDate,
    private val reviews: ReviewList,
    private val tasks: TaskList,
    private val areas: AreaList,
    private val goals: GoalList,
    private val habits: HabitList,
    private val prompts: PromptLibrary,
    private val rituals: RitualRunList,
    private val io: CoroutineDispatcher,
    private val today: () -> LocalDate,
) : ViewModel() {
    private val step = MutableStateFlow(ReviewStep.LOOK_BACK)
    private val questions = MutableStateFlow<List<ReviewQuestion>>(emptyList())
    private val answers = MutableStateFlow<Map<String, String>>(emptyMap())
    private val opened = MutableStateFlow<ReviewItem?>(null)

    val uiState: StateFlow<ReviewUiState> = combine(
        reviews.watch().flowOn(io),
        tasks.watchAll().flowOn(io),
        goals.watch().flowOn(io),
        habits.watch().flowOn(io),
        combine(step, questions, answers, opened) { current, asked, written, review -> Screen(current, asked, written, review) },
    ) { allReviews, taskList, (goalList, entries), habitData, screen ->
        val day = today()
        val digest = ReviewLookBack.build(kind, periodStart, taskList, areas.all(), goalList, entries, habitData, day)
        val nextStart = ReviewLookBack.periodEnd(kind, periodStart).plusDays(1)
        val horizon = ReviewLookBack.horizonOf(kind)
        val board = GoalBoard.build(goalList, entries, taskList, day, habits = habitData)
        val next = board.sections.firstOrNull { it.horizon == horizon && it.start == nextStart }
        ReviewUiState(
            loaded = true,
            step = screen.step,
            kind = kind,
            digest = digest,
            review = allReviews.firstOrNull { it.kind == kind && it.periodStart == periodStart } ?: screen.review,
            questions = screen.questions,
            answers = screen.answers,
            nextGoals = next?.rows.orEmpty(),
            canCopyGoals = next?.canCopy ?: goalList.none { it.horizon == horizon && it.periodStart == nextStart },
            goals = goalList,
        )
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), ReviewUiState(kind = kind))

    init {
        viewModelScope.launch(io) {
            val review = reviews.open(kind, periodStart)
            opened.value = review
            answers.value = review?.reflections.orEmpty().associate { it.promptId to it.answer }
            questions.value = ask(review)
        }
    }

    /** The prompts this review asks: the ones the period calls for, then the rotation (docs/reviews.md). */
    private fun ask(review: ReviewItem?): List<ReviewQuestion> {
        val already = review?.reflections.orEmpty()
        if (already.isNotEmpty()) {
            return already.map { reflection ->
                val prompt = prompts[reflection.promptId]
                ReviewQuestion(
                    reflection.promptId,
                    prompt?.let { PromptRules.text(it, kind) } ?: reflection.promptId,
                )
            }
        }
        val digest = ReviewLookBack.build(kind, periodStart, tasks.all(), areas.all(), goals.all(), goals.entries(), habits.read(), today())
        val reactive = PromptRules.reactive(prompts, kind, digest.facts).take(MAX_REACTIVE)
        // What past reviews of this kind already asked, oldest first, so the rotation moves on.
        val shown = reviews.all().filter { it.kind == kind }.sortedBy { it.periodStart }.flatMap { it.reflections }.map { it.promptId }
        val rotation = PromptRules.rotation(prompts, kind, shown, QUESTIONS - reactive.size)
        return reactive + rotation.map { prompt -> ReviewQuestion(prompt.id, PromptRules.text(prompt, kind)) }
    }

    /** Writes an answer; it is saved when the step is left, and at once when the review is finished. */
    fun answer(promptId: String, text: String) = answers.update { it + (promptId to text) }

    fun setMood(mood: Int) = write { review -> reviews.setMood(review.id, mood.takeIf { it != uiState.value.mood }) }

    fun setEnergy(energy: Int) = write { review -> reviews.setEnergy(review.id, energy.takeIf { it != uiState.value.energy }) }

    /** What to do with a task still open from the period: to the next period, done, or dropped. */
    fun decide(task: TaskItem, decision: PlanDecision) {
        viewModelScope.launch(io) {
            when (decision) {
                PlanDecision.TOMORROW -> tasks.plan(task.id, maxOf(today(), ReviewLookBack.periodEnd(kind, periodStart).plusDays(1)))
                PlanDecision.DONE -> tasks.setDone(task.id, true)
                PlanDecision.DROPPED -> tasks.drop(task.id)
                else -> Unit
            }
        }
    }

    /** Copies the goals of this period into the next one, when it has none yet (docs/goals.md). */
    fun copyGoals() {
        viewModelScope.launch(io) {
            goals.copyPrevious(ReviewLookBack.horizonOf(kind), ReviewLookBack.periodEnd(kind, periodStart).plusDays(1))
        }
    }

    /** Adds a goal to the period ahead. False when the draft isn't one the goal list keeps. */
    suspend fun addGoal(draft: GoalDraft): Boolean = withContext(io) { goals.add(draft) != null }

    /** The goal the Add button starts from: the next period, of this review's horizon. */
    fun nextGoalDraft(): GoalDraft = GoalDraft(
        title = "",
        horizon = ReviewLookBack.horizonOf(kind),
        periodStart = ReviewLookBack.periodEnd(kind, periodStart).plusDays(1),
        mode = GoalRules.MODE_DONE,
    )

    fun next() {
        saveAnswers()
        step.value = when (step.value) {
            ReviewStep.LOOK_BACK -> ReviewStep.TASKS
            ReviewStep.TASKS -> ReviewStep.REFLECT
            ReviewStep.REFLECT -> ReviewStep.RATE
            ReviewStep.RATE -> ReviewStep.GOALS
            ReviewStep.GOALS, ReviewStep.DONE -> {
                viewModelScope.launch(io) { rituals.record(ritual(), today()) }
                ReviewStep.DONE
            }
        }
    }

    /** Back a step; false where back should leave the review. */
    fun back(): Boolean {
        val before = step.value
        if (before == ReviewStep.LOOK_BACK) return false
        saveAnswers()
        step.value = when (before) {
            ReviewStep.TASKS -> ReviewStep.LOOK_BACK
            ReviewStep.REFLECT -> ReviewStep.TASKS
            ReviewStep.RATE -> ReviewStep.REFLECT
            ReviewStep.GOALS -> ReviewStep.RATE
            else -> ReviewStep.GOALS
        }
        return true
    }

    /** Keeps what was written; the screen calls it when it leaves, so nothing is lost. */
    fun saveAnswers() {
        val review = uiState.value.review ?: opened.value ?: return
        val asked = questions.value
        if (asked.isEmpty()) return
        val written = answers.value
        viewModelScope.launch(io) {
            reviews.setReflections(review.id, asked.map { Reflection(it.promptId, written[it.promptId].orEmpty()) })
        }
    }

    private fun ritual(): String = if (kind == "monthly") RitualRunList.MONTHLY_REVIEW else RitualRunList.WEEKLY_REVIEW

    private fun write(work: (ReviewItem) -> Unit) {
        val review = uiState.value.review ?: opened.value ?: return
        viewModelScope.launch(io) { work(review) }
    }

    private data class Screen(
        val step: ReviewStep,
        val questions: List<ReviewQuestion>,
        val answers: Map<String, String>,
        val review: ReviewItem?,
    )

    private companion object {
        /** How many prompts a review asks, and how many of them may come from the period's facts. */
        const val QUESTIONS = 3
        const val MAX_REACTIVE = 2
    }
}

File: windows/src/GoalMaker.App/ViewModels/ReviewViewModel.cs
using System.Collections.ObjectModel;
using System.Globalization;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using GoalMaker.App.Localization;
using GoalMaker.Core.Planning;
using GoalMaker.Core.Settings;

namespace GoalMaker.App.ViewModels;

/// <summary>
/// The guided review of one period (docs/reviews.md, spec stories 59 to 64): look back, handle what is
/// still open, reflect on the prompts the library and the period's facts give, rate mood and energy,
/// and set the next period's goals. Everything is saved as it is answered.
/// </summary>
public sealed partial class ReviewViewModel : ObservableObject
{
    // How many prompts a review asks, and how many of them may come from the period's facts.
    private const int Questions = 3;
    private const int MaxReactive = 2;

    private readonly ReviewList reviews;
    private readonly TaskList tasks;
    private readonly AreaList areas;
    private readonly GoalList goals;
    private readonly HabitList habits;
    private readonly PromptLibrary prompts;
    private readonly RitualRunList rituals;
    private readonly ISettingsStore settings;
    private readonly IStrings strings;
    private readonly TimeProvider time;
    private ReviewItem? review;
    private bool loading;

    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(IsLookBack), nameof(IsTasks), nameof(IsReflect), nameof(IsRate), nameof(IsGoals), nameof(IsDone), nameof(NextLabel), nameof(Progress), nameof(CanGoBack))]
    private ReviewStep step = ReviewStep.LookBack;

    [ObservableProperty]
    private ReviewDigest digest = new();

    [ObservableProperty]
    private string heading = string.Empty;

    [ObservableProperty]
    private string periodText = string.Empty;

    [ObservableProperty]
    private string doneText = string.Empty;

    [ObservableProperty]
    private string changeText = string.Empty;

    [ObservableProperty]
    private IReadOnlyList<string> highlights = [];

    [ObservableProperty]
    private int? mood;

    [ObservableProperty]
    private int? energy;

    [ObservableProperty]
    private string summaryText = string.Empty;

    public ReviewViewModel(
        string kind,
        DateOnly periodStart,
        ReviewList reviews,
        TaskList tasks,
        AreaList areas,
        GoalList goals,
        HabitList habits,
        PromptLibrary prompts,
        RitualRunList rituals,
        ISettingsStore settings,
        IStrings strings,
        TimeProvider time,
        Action<Action> runOnUi)
    {
        Kind = kind;
        PeriodStart = periodStart;
        this.reviews = reviews;
        this.tasks = tasks;
        this.areas = areas;
        this.goals = goals;
        this.habits = habits;
        this.prompts = prompts;
        this.rituals = rituals;
        this.settings = settings;
        this.strings = strings;
        this.time = time;
        tasks.Changed += (_, _) => runOnUi(Refresh);
        goals.Changed += (_, _) => runOnUi(Refresh);
        habits.Changed += (_, _) => runOnUi(Refresh);
        Start();
    }

    public string Kind { get; private set; }

    public DateOnly PeriodStart { get; private set; }

    public ObservableCollection<ReviewQuestionViewModel> QuestionRows { get; } = [];

    public ObservableCollection<ReviewTaskViewModel> OpenTasks { get; } = [];

    public ObservableCollection<GoalRowViewModel> Goals { get; } = [];

    public ObservableCollection<ReviewDigest.Day> Days { get; } = [];

    public ObservableCollection<ReviewDigest.Goal> PeriodGoals { get; } = [];

    public ObservableCollection<ReviewDigest.Habit> PeriodHabits { get; } = [];

    public bool IsLookBack => Step == ReviewStep.LookBack;

    public bool IsTasks => Step == ReviewStep.Tasks;

    public bool IsReflect => Step == ReviewStep.Reflect;

    public bool IsRate => Step == ReviewStep.Rate;

    public bool IsGoals => Step == ReviewStep.Goals;

    public bool IsDone => Step == ReviewStep.Done;

    /// <summary>How far the review has come, 0 to 1, for the bar across the top.</summary>
    public double Progress => ((int)Step + 1) / (double)Enum.GetValues<ReviewStep>().Length;

    public string NextLabel => strings.Get(Step switch
    {
        ReviewStep.Goals => "Reviews.Finish",
        ReviewStep.Done => "Reviews.Close",
        _ => "Reviews.Next",
    });

    /// <summary>Opens (or picks up) the review of a period; the page reuses one view model.</summary>
    public void Open(string kind, DateOnly periodStart)
    {
        Kind = kind;
        PeriodStart = periodStart;
        Step = ReviewStep.LookBack;
        Start();
    }

    public void Refresh()
    {
        var today = Today();
        Digest = ReviewLookBack.Build(
            Kind,
            PeriodStart,
            tasks.All(),
            areas.All(),
            goals.All(),
            goals.Entries(),
            habits.All(),
            habits.Checkins(),
            habits.Pauses(),
            today);
        Heading = strings.Get(Kind == ReviewRules.Monthly ? "Reviews.Monthly" : "Reviews.Weekly");
        PeriodText = strings.Get(
            "Goals.Range",
            Digest.PeriodStart.ToString("d MMM", CultureInfo.CurrentCulture),
            Digest.PeriodEnd.ToString("d MMM", CultureInfo.CurrentCulture));
        DoneText = strings.Get("Reviews.Done", Digest.Done);
        ChangeText = Digest.DoneBefore == 0
            ? strings.Get("Reviews.NoLastPeriod")
            : Digest.Change > 0
                ? strings.Get("Reviews.MoreThanBefore", Digest.Change)
                : Digest.Change < 0
                    ? strings.Get("Reviews.FewerThanBefore", -Digest.Change)
                    : strings.Get("Reviews.SameAsBefore");
        var highlights = new List<string>();
        if (Digest.BestDay is { } best)
        {
            highlights.Add(strings.Get("Reviews.BestDay", best.Date.ToString("dddd", CultureInfo.CurrentCulture), best.Done));
        }

        if (Digest.StrongestArea is { } area)
        {
            highlights.Add(strings.Get("Reviews.StrongestArea", area.Name, area.Done));
        }

        if (Digest.Habits.OrderByDescending(habit => habit.Streak).FirstOrDefault() is { Streak: > 0 } streak)
        {
            highlights.Add(strings.Get("Reviews.LongestStreak", streak.Name, streak.Streak));
        }

        Highlights = highlights;
        Days.Clear();
        foreach (var day in Digest.Days)
        {
            Days.Add(day);
        }

        PeriodGoals.Clear();
        foreach (var goal in Digest.Goals)
        {
            PeriodGoals.Add(goal);
        }

        PeriodHabits.Clear();
        foreach (var habit in Digest.Habits)
        {
            PeriodHabits.Add(habit);
        }

        OpenTasks.Clear();
        foreach (var task in Digest.OpenTasks)
        {
            OpenTasks.Add(new ReviewTaskViewModel(task, Decide));
        }

        ShowNextGoals();
        OnPropertyChanged(nameof(HasOpenTasks));
        OnPropertyChanged(nameof(HasPeriodGoals));
        OnPropertyChanged(nameof(HasPeriodHabits));
        OnPropertyChanged(nameof(MostDone));
        OnPropertyChanged(nameof(OpenTasksText));
        OnPropertyChanged(nameof(DoneSummary));
    }

    public bool HasOpenTasks => OpenTasks.Count > 0;

    public bool HasPeriodGoals => PeriodGoals.Count > 0;

    public bool HasPeriodHabits => PeriodHabits.Count > 0;

    /// <summary>The busiest day of the period, which fills a day's bar.</summary>
    public int MostDone => Days.Count == 0 ? 1 : Math.Max(1, Days.Max(day => day.Done));

    /// <summary>Back leaves the first step to the page, not to the review.</summary>
    public bool CanGoBack => Step != ReviewStep.LookBack;

    public string OpenTasksText => strings.Get(HasOpenTasks ? "Reviews.OpenIntro" : "Reviews.OpenNone");

    public string DoneSummary => strings.Get("Reviews.DoneSummary", Digest.Done, QuestionRows.Count(question => !string.IsNullOrWhiteSpace(question.Answer)));

    public bool CanCopyGoals { get; private set; }

    /// <summary>Keeps what was written; the page calls it when it leaves, so nothing is lost.</summary>
    public void SaveAnswers()
    {
        if (review is null || QuestionRows.Count == 0 || loading)
        {
            return;
        }

        reviews.SetReflections(review.Id, QuestionRows.Select(question => new Reflection(question.PromptId, question.Answer)));
    }

    [RelayCommand]
    private void Next()
    {
        SaveAnswers();
        if (Step == ReviewStep.Goals)
        {
            rituals.Record(Kind == ReviewRules.Monthly ? RitualRunList.MonthlyReview : RitualRunList.WeeklyReview, Today());
            Step = ReviewStep.Done;
            OnPropertyChanged(nameof(DoneSummary));
            return;
        }

        if (Step != ReviewStep.Done)
        {
            Step = (ReviewStep)((int)Step + 1);
        }
    }

    [RelayCommand]
    private void Back()
    {
        if (Step == ReviewStep.LookBack)
        {
            return;
        }

        SaveAnswers();
        Step = (ReviewStep)((int)Step - 1);
    }

    [RelayCommand]
    private void SetMood(int value)
    {
        if (review is null)
        {
            return;
        }

        Mood = Mood == value ? null : value;
        reviews.SetMood(review.Id, Mood);
    }

    [RelayCommand]
    private void SetEnergy(int value)
    {
        if (review is null)
        {
            return;
        }

        Energy = Energy == value ? null : value;
        reviews.SetEnergy(review.Id, Energy);
    }

    [RelayCommand]
    private void CopyGoals()
    {
        goals.CopyPrevious(ReviewLookBack.HorizonOf(Kind), ReviewLookBack.PeriodEnd(Kind, PeriodStart).AddDays(1));
        ShowNextGoals();
    }

    private void Start()
    {
        loading = true;
        try
        {
            review = reviews.Open(Kind, PeriodStart);
            Mood = review?.Mood;
            Energy = review?.Energy;
            SummaryText = review?.Summary ?? string.Empty;
            Refresh();
            ShowQuestions();
        }
        finally
        {
            loading = false;
        }
    }

    // The prompts this review asks: the ones it asked before, or the period's own plus the rotation.
    private void ShowQuestions()
    {
        QuestionRows.Clear();
        var asked = review?.Reflections ?? [];
        if (asked.Count > 0)
        {
            foreach (var reflection in asked)
            {
                var prompt = prompts[reflection.PromptId];
                QuestionRows.Add(new ReviewQuestionViewModel(
                    reflection.PromptId,
                    prompt is null ? reflection.PromptId : PromptRules.Text(prompt, Kind),
                    reflection.Answer,
                    SaveAnswers));
            }

            return;
        }

        var reactive = PromptRules.Reactive(prompts, Kind, Digest.Facts).Take(MaxReactive).ToList();
        var shown = reviews.All()
            .Where(other => other.Kind == Kind)
            .OrderBy(other => other.PeriodStart)
            .SelectMany(other => other.Reflections)
            .Select(reflection => reflection.PromptId)
            .ToList();
        foreach (var question in reactive)
        {
            QuestionRows.Add(new ReviewQuestionViewModel(question.PromptId, question.Text, string.Empty, SaveAnswers));
        }

        foreach (var prompt in PromptRules.Rotation(prompts, Kind, shown, Questions - reactive.Count))
        {
            QuestionRows.Add(new ReviewQuestionViewModel(prompt.Id, PromptRules.Text(prompt, Kind), string.Empty, SaveAnswers));
        }
    }

    private void ShowNextGoals()
    {
        var horizon = ReviewLookBack.HorizonOf(Kind);
        var nextStart = ReviewLookBack.PeriodEnd(Kind, PeriodStart).AddDays(1);
        var all = goals.All();
        var own = all.Where(goal => goal.Horizon == horizon && goal.PeriodStart == nextStart && goal.Status != GoalRules.Dropped).ToList();
        var entries = goals.Entries().ToLookup(entry => entry.GoalId, StringComparer.Ordinal);
        var taskList = tasks.All().Where(task => task.GoalId is not null).ToLookup(task => task.GoalId!, StringComparer.Ordinal);
        Goals.Clear();
        foreach (var goal in own)
        {
            var progress = GoalRules.Progress(goal.Mode, goal.Status, goal.Target, taskList[goal.Id], entries[goal.Id]);
            Goals.Add(new GoalRowViewModel(goal, progress, null, 0, strings));
        }

        CanCopyGoals = own.Count == 0 && all.Any(goal => goal.Horizon == horizon && goal.PeriodStart == PeriodStart && goal.Status != GoalRules.Dropped);
        OnPropertyChanged(nameof(CanCopyGoals));
    }

    private void Decide(TaskItem task, string decision)
    {
        switch (decision)
        {
            case "forward":
                var next = ReviewLookBack.PeriodEnd(Kind, PeriodStart).AddDays(1);
                tasks.Plan(task.Id, next > Today() ? next : Today());
                break;
            case "done":
                tasks.SetDone(task.Id, true);
                break;
            default:
                tasks.Drop(task.Id);
                break;
        }
    }

    private DateOnly Today() => PlanningDay.Of(time.GetLocalNow().DateTime, settings.DayStartHour);
}
